File: 2025/day12/main.py
*S, I = open(0).read().split('\n\n')
szs = []
shs = []
for s in S:
    szs.append(s.count('#'))
    ls = s.splitlines()[1:]
    sh = [[int(c == '#') for c in l] for l in ls]
    hs = set()
    for _ in range(4):
        hs.add(tuple(sum(b << j for j, b in enumerate(row)) for row in sh))
        hs.add(tuple(sum(b << (2 - j) for j, b in enumerate(row)) for row in sh))
        sh = [[sh[j][2 - i] for j in range(3)] for i in range(3)]
    shs.append(list(hs))

def f(g, n, m, xs, r=0, c=0):
    if sum(xs) == 0:
        return True
    for r in range(r, n - 2):
        temp = g[r:r+3]
        for c in range(c, m - 2):
            for i, x in enumerate(xs):
                if x == 0:
                    continue
                for h in shs[i]:
                    if any((g[r + j] >> c) & h[j] for j in range(3)):
                        continue
                    xs[i] -= 1
                    for j, b in enumerate(h):
                        g[r + j] |= b << c
                    if f(g, n, m, xs, r, c + 1):
                        return True
                    g[r:r+3] = temp
                    xs[i] += 1
        c = 0
    return False

ans = 0
for l in I.splitlines():
    dims, *a = l.split()
    n, m = map(int, dims[:-1].split('x'))
    xs = [int(x) for x in a]
    t = n * m
    if t < sum(sz * x for sz, x in zip(szs, xs)):
        continue
    g = [0] * n
    ans += f(g, n, m, xs)
print(ans)

File: 2025/day12/src/main.rs
use std::io::{self, Read};

const SHAPE_DIM: usize = 3;
const FILLED_CELL: u8 = b'#';

fn rotate_grid<T: Copy>(grid: &[Vec<T>]) -> Vec<Vec<T>> {
    let n = grid.len();
    Vec::from_iter((0..n).map(|i| Vec::from_iter((0..n).map(|j| grid[j][n - 1 - i]))))
}

struct Shape {
    area: u32,
    masks: Vec<[u64; SHAPE_DIM]>,
}

impl Shape {
    fn parse(input: &str) -> Self {
        let area = input
            .as_bytes()
            .iter()
            .filter(|&&b| b == FILLED_CELL)
            .count() as u32;
        let mut grid: Vec<_> = input
            .lines()
            .skip(1)
            .map(|row| row.as_bytes().to_vec())
            .collect();
        let mut masks = Vec::new();
        for _ in 0..4 {
            let mut mask1 = [0; SHAPE_DIM];
            let mut mask2 = [0; SHAPE_DIM];
            for (i, row) in grid.iter().enumerate() {
                mask1[i] = row
                    .iter()
                    .fold(0, |acc, &cell| (acc << 1) | u64::from(cell == FILLED_CELL));
                mask2[i] = row
                    .iter()
                    .rfold(0, |acc, &cell| (acc << 1) | u64::from(cell == FILLED_CELL));
            }
            masks.push(mask1);
            masks.push(mask2);
            grid = rotate_grid(&grid);
        }
        masks.sort_unstable();
        masks.dedup();
        Shape { area, masks }
    }
}

struct Query {
    width: usize,
    height: usize,
    shape_counts: Vec<u32>,
}

impl Query {
    fn parse(input: &str) -> Option<Self> {
        let mut iter = input.split_ascii_whitespace();
        let (width, height) = iter.next()?.strip_suffix(':')?.split_once('x')?;
        let shape_counts: Vec<_> = iter.map(|x| x.parse().unwrap()).collect();
        Some(Query {
            width: width.parse().ok()?,
            height: height.parse().ok()?,
            shape_counts,
        })
    }
}

fn can_fit(shapes: &[Shape], query: &Query) -> bool {
    fn helper(
        shapes: &[Shape],
        width: usize,
        height: usize,
        shape_counts: &mut [u32],
        grid: &mut [u64],
        start_row: usize,
        start_col: usize,
    ) -> bool {
        if shape_counts.iter().all(|&count| count == 0) {
            return true;
        }
        for r in start_row..height - SHAPE_DIM + 1 {
            let start_c = if r == start_row { start_col } else { 0 };
            for c in start_c..width - SHAPE_DIM + 1 {
                for (i, shape) in shapes.iter().enumerate() {
                    if shape_counts[i] == 0 {
                        continue;
                    }
                    shape_counts[i] -= 1;
                    for shape_mask in &shape.masks {
                        let shape_blocked = grid[r..r + SHAPE_DIM]
                            .iter()
                            .zip(shape_mask)
                            .any(|(&row_mask, m)| row_mask & (m << c) != 0);
                        if shape_blocked {
                            continue;
                        }
                        for (row_mask, m) in grid[r..r + SHAPE_DIM].iter_mut().zip(shape_mask) {
                            *row_mask |= m << c;
                        }
                        if helper(shapes, width, height, shape_counts, grid, r, c + 1) {
                            return true;
                        }
                        for (row_mask, m) in grid[r..r + SHAPE_DIM].iter_mut().zip(shape_mask) {
                            *row_mask ^= m << c;
                        }
                    }
                    shape_counts[i] += 1;
                }
            }
        }
        false
    }
    helper(
        shapes,
        query.width,
        query.height,
        &mut query.shape_counts.clone(),
        &mut vec![0; query.height],
        0,
        0,
    )
}

fn main() -> io::Result<()> {
    let stdin = io::stdin();
    let mut input = String::new();
    stdin.lock().read_to_string(&mut input)?;
    let mut input_parts: Vec<_> = input.split("\n\n").collect();
    let queries = input_parts.pop().unwrap();
    let shapes: Vec<_> = input_parts.into_iter().map(Shape::parse).collect();

    let mut ans = 0;
    for query in queries.lines().map(Query::parse) {
        let query = query.unwrap();
        let total_shape_area = query
            .shape_counts
            .iter()
            .zip(&shapes)
            .fold(0, |acc, (&count, shape)| acc + count * shape.area);
        if ((query.width * query.height) as u32) < total_shape_area {
            continue;
        }
        let total_shapes = query.shape_counts.iter().sum();
        if ((query.width / SHAPE_DIM) * (query.height / SHAPE_DIM)) as u32 >= total_shapes {
            ans += 1;
            continue;
        }
        ans += u32::from(can_fit(&shapes, &query));
    }

    println!("{}", ans);
    Ok(())
}
